### PR TITLE
[location-component] Add tests to improve code coverage.

### PR DESCRIPTION
**Pull request by pengdev**
PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).
Fixes: #68 
## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [x] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog>Improve test coverage of location component plugin.</changelog>`.

### Summary of changes
This PR adds unit tests for the location component plugin, it also fixes the crash when location plugin puck setting changes during the plugin is  disabled and enabled.

### CODE REVIEW

File: plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    locationComponentPlugin.removeOnIndicatorBearingChangedListener(bearingListener)
+    locationComponentPlugin.indicatorBearingChangedListener.onIndicatorBearingChanged(0.0)
+    verify(exactly = 0) { bearingListener.onIndicatorBearingChanged(0.0) }
+  }
+
+  @Test
+  fun testSetLocationProvider() {
+    val mockLocationProvider = mockk<LocationProvider>(relaxed = true)
+    preparePluginInitialisationWithEnabled()
+    verify { locationProvider.registerLocationConsumer(any()) }
```
Comment (kiryldz) [inline plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt:183]: Verify it happens only once?

Comment (kiryldz) [reply plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt:183]: Same to other similar verifies.

File: plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+
+  @Test
+  fun testOnStyleChanged() {
+    every { locationPuckManager.isLayerInitialised() } returns false
+    preparePluginInitialisationWithEnabled()
+    verify(exactly = 2) { locationPuckManager.isLayerInitialised() }
+    verify(exactly = 1) { locationPuckManager.initialize(style) }
+    locationComponentPlugin.onStyleChanged(style)
+    verify(exactly = 3) { locationPuckManager.isLayerInitialised() }
+    verify(exactly = 2) { locationPuckManager.initialize(style) }
```
Comment (kiryldz) [inline plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt:208]: So it's expected to initialize `locationPuckManager` twice here?

File: plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    every { delegateProvider.getStyle(capture(styleCallbackSlot)) } returns Unit
+    locationComponentPlugin.enabled = true
+    if (styleCallbackSlot.isCaptured) {
+      styleCallbackSlot.captured.invoke(style)
+    }
+    verify(exactly = 1) { style.addStyleLayer(any(), any()) }
+    verify(exactly = 2) { locationProvider.registerLocationConsumer(any()) }
+    verify(exactly = 1) { locationProvider.unRegisterLocationConsumer(any()) }
+  }
+}
```
Comment (kiryldz) [inline plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt:354]: It's pretty hard to check if all cases are covered but can we have a test (or did I miss it) to check that recently added condition https://github.com/mapbox/mapbox-maps-android/blob/c62398a7ea11e0c5ac99dde69962baaa12b78827/plugin-locationcomponent/src/main/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImpl.kt#L133

File: plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+  }
+
+  @Test
+  fun testOnStyleChangedWhilePuckManagerInitialised() {
+    every { locationPuckManager.isLayerInitialised() } returns true
+    preparePluginInitialisationWithEnabled()
+    verify(exactly = 2) { locationPuckManager.isLayerInitialised() }
+    verify(exactly = 0) { locationPuckManager.initialize(style) }
+    locationComponentPlugin.onStyleChanged(style)
+    verify(exactly = 3) { locationPuckManager.isLayerInitialised() }
```
Comment (kiryldz) [inline plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt:218]: Guess it's not that useful to check call count of getter function.
We are more interested of how many times `initialize` was called.

File: plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt (hunk 2 above)
Comment (pengdev) [reply plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt:208]: The `fun preparePluginInitialisationWithEnabled` has already initialised the puck manager once, when style changes, we need to initialise the location manager again.

File: plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt (hunk 4 above)
Comment (pengdev) [reply plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt:218]: In the unit tests we verify all the calls to the locationPuckManager, so I guess it's fine to leave it.

File: plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt (hunk 3 above)
Comment (pengdev) [reply plugin-locationcomponent/src/test/java/com/mapbox/maps/plugin/locationcomponent/LocationComponentPluginImplTest.kt:354]: It's covered in `testOnStart`, `testOnStartWithPuckDisabled` and `testOnStartWithPuckAlreadyInitialisedAndPluginActivated`.

**Review by kiryldz — APPROVED** [review]
@pengdev great job!
Thanks for clarifying several questions, LGTM!

**Pull request merged**